group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1197,6 +1197,45 @@ private void throwIfClassicProtocolIsNotSupported(
         }
     }
 
+    /**
+     * Validates if the consumer group member uses the classic protocol.
+     *
+     * @param member The ConsumerGroupMember.
+     */
+    private void throwIfMemberDoesNotUseClassicProtocol(ConsumerGroupMember member) {
+        if (!member.useClassicProtocol()) {
+            throw new UnknownMemberIdException(
+                String.format("Member %s does not use the classic protocol.", member.memberId())
+            );
+        }
+    }
+
+    /**
+     * Validates if the generation id and the protocol type from the request match those of the consumer group.
+     *
+     * @param group                 The ConsumerGroup.
+     * @param member                The ConsumerGroupMember.
+     * @param requestGenerationId   The generation id from the request.
+     * @param requestProtocolType   The protocol type from the request.
+     * @param requestProtocolName   The protocol name from the request.
+     */
+    private void throwIfGenerationIdOrProtocolUnmatched(
+        ConsumerGroup group,
+        ConsumerGroupMember member,
+        int requestGenerationId,
+        String requestProtocolType,
+        String requestProtocolName
+    ) {
+        if (member.memberEpoch() != requestGenerationId) {
+            throw Errors.ILLEGAL_GENERATION.exception(
+                String.format("The request generation id %s is not equal to the group epoch %d from the consumer group %s.",
+                    requestGenerationId, group.groupEpoch(), group.groupId())
+            );
+        } else if (!group.supportsClassicProtocols(requestProtocolType, Collections.singleton(requestProtocolName))) {
+            throw Errors.INCONSISTENT_GROUP_PROTOCOL.exception("The member protocol is not supported.");
+        }
```

Comment:
This is incorrect. In the JoinGroup response, we set the protocol name to `protocols.iterator().next().name()`. Here we should validate that we get it back.